### PR TITLE
CART-922

### PR DESCRIPTION
**Pull request by aosinskx**

### CODE REVIEW

File: src/gurt/dyn_hash.c [hunk 1]
```diff
@@ -0,0 +1,1188 @@
...
+#include <gurt/dyn_hash.h>
+
+#define DYNHASH_SIPBITS 6
+#define DYNHASH_BUCKET	(1 << DYNHASH_SIPBITS)
+#define DYNHASH_VECTOR	64
+#define DYNHASH_MAGIC   0xab013245
+
+#define _le64toh(x) (x)
+
+typedef struct dh_field {
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/dyn_hash.c:39]: Ignore

File: src/gurt/dyn_hash.c [hunk 2]
```diff
@@ -0,0 +1,1188 @@
...
+typedef struct dh_bucket {
+	unsigned char	counter;
+	pthread_mutex_t	mtx;
+	dh_field_t	field[DYNHASH_BUCKET];
+} dh_bucket_t;
+
+
+#define ROTATE(x, b) (uint64_t)(((x) << (b)) | ((x) >> (64 - (b))))
+
+#define HALF_ROUND(a, b, c, d, s, t) \
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/dyn_hash.c:53]: Ignore

File: src/gurt/dyn_hash.c [hunk 3]
```diff
@@ -0,0 +1,1188 @@
...
+    a += b; c += d;                  \
+    b = ROTATE(b, s) ^ a;            \
+    d = ROTATE(d, t) ^ c;            \
+    a = ROTATE(a, 32)
+
+#define DOUBLE_ROUND(v0, v1, v2, v3)         \
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/dyn_hash.c:59]: Ignore

File: src/gurt/tests/test_gurt.c [hunk 4]
```diff
@@ -1028,6 +1038,7 @@ static d_hash_table_ops_t th_ops_ref = {
 	.hop_rec_decref		= test_gurt_hash_op_rec_decref,
 	.hop_rec_ndecref	= test_gurt_hash_op_rec_ndecref,
 	.hop_rec_free		= test_gurt_hash_op_rec_free,
+	.hop_key_get		= test_gurt_hash_op_key_get,
```
Comment (dmiter) [inline src/gurt/tests/test_gurt.c:1041]: I removed this function a few time ago. Why you use old version? Please rebase on top of current master.

**Review by dmiter — CHANGES_REQUESTED** [review]
Please read "https://wiki.hpdd.intel.com/display/DC/C+Style+Guide" and use this coding style.
It's very difficult to read if you use wrong coding style and especially alignment.

It looks you based on very old version of code. Please rebase it on top of fresh master.

File: src/gurt/hash.c [hunk 5]
```diff
@@ -443,7 +448,7 @@ d_hash_rec_find(struct d_hash_table *htable, const void *key,
 		ch_rec_addref(htable, link);
 
 	ch_bucket_unlock(htable, idx, !is_lru);
-	return link;
+	return (void *)link;
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:451]: Yes, it's a typo, thank you.

File: src/gurt/tests/test_gurt.c (hunk 4 above)
Comment (aosinskx) [reply src/gurt/tests/test_gurt.c:1041]: This function is necessary for dynamic hash in order to associate to find object. In the existing hash, the object type is d_list_t* and could be removed from the hash directly. However, in the dynamic hash, the key is necessary to find the object if d_hash_rec_decref(),  d_hash_rec_ndecref(), forces object to get removed.

File: src/gurt/hash.c [hunk 6]
```diff
@@ -452,8 +457,13 @@ d_hash_rec_insert(struct d_hash_table *htable, const void *key,
...
 	struct d_hash_bucket	*bucket;
 	d_list_t		*tmp;
-	uint32_t		 idx;
-	int			 rc = 0;
+	uint32_t		idx;
+	int			rc = 0;
+
+	if (htable->ht_feats & D_HASH_FT_DYNAMIC) {
+		return dyn_hash_rec_insert(htable, key, ksize,
+		(void *)link, exclusive);
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:465]: Done.

File: src/gurt/hash.c [hunk 7]
```diff
@@ -473,14 +483,19 @@ d_hash_rec_insert(struct d_hash_table *htable, const void *key,
...
 d_hash_rec_find_insert(struct d_hash_table *htable, const void *key,
 		       unsigned int ksize, d_list_t *link)
 {
 	struct d_hash_bucket	*bucket;
 	d_list_t		*tmp;
 	uint32_t		 idx;
 
+	if (htable->ht_feats & D_HASH_FT_DYNAMIC) {
+		return (d_list_t *)dyn_hash_rec_find_insert(htable, key,
+		ksize, (void *)link, 0);
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:496]: Done.

File: src/gurt/hash.c [hunk 8]
```diff
@@ -512,6 +527,10 @@ d_hash_rec_insert_anonym(struct d_hash_table *htable, d_list_t *link,
 	if (htable->ht_ops->hop_key_init == NULL)
 		return -DER_INVAL;
 
+	if (htable->ht_feats & D_HASH_FT_DYNAMIC) {
+		return -DER_NOTAPPLICABLE;
+	}
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:532]: Done

File: src/gurt/hash.c [hunk 9]
```diff
@@ -548,6 +567,10 @@ d_hash_rec_delete(struct d_hash_table *htable, const void *key,
 	bool			 deleted = false;
 	bool			 zombie  = false;
 
+	if (htable->ht_feats & D_HASH_FT_DYNAMIC) {
+		return dyn_hash_rec_delete(htable, key, ksize, 0);
+	}
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:572]: Done.

File: src/gurt/hash.c [hunk 10]
```diff
@@ -601,8 +628,10 @@ d_hash_rec_evict(struct d_hash_table *htable, const void *key,
 	d_list_t		*link;
 	uint32_t		 idx;
 
-	if (!(htable->ht_feats & D_HASH_FT_LRU))
+	if (!(htable->ht_feats & D_HASH_FT_LRU) ||
+			(htable->ht_feats & D_HASH_FT_DYNAMIC)) {
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:632]: Done

File: src/gurt/hash.c [hunk 11]
```diff
@@ -761,10 +810,14 @@ d_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
...
-	uint32_t nr = (1 << bits);
-	uint32_t i;
-	int	 rc = 0;
+	uint32_t	nr = (1 << bits);
+	uint32_t	i;
+	int		rc = 0;
 
+	if (feats & D_HASH_FT_DYNAMIC) {
+		return dyn_hash_table_create_inplace(feats, bits,
+		priv, hops, htable);
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/hash.c:819]: Why are you surprised? It's autopilot - once done, repeated multiple times.
Fixed.

File: src/gurt/dyn_hash.c [hunk 12]
```diff
@@ -739,6 +771,7 @@ dyn_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
 		} else if (feats & D_HASH_FT_RWLOCK) {
 			rc = D_RWLOCK_INIT(&gtable->ht_lock.rwlock, NULL);
 			if (rc != 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:773]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 13]
```diff
@@ -748,6 +781,7 @@ dyn_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
 			rc = D_SPIN_INIT(&gtable->ht_lock.spin,
 					 PTHREAD_PROCESS_PRIVATE);
 			if (rc != 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:783]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 14]
```diff
@@ -759,20 +793,23 @@ dyn_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
 	/* initialize vector */
 	rc = vec_init(&htable->ht_vector, DYNHASH_SIPBITS);
 	if (rc != 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:795]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 15]
```diff
@@ -759,20 +793,23 @@ dyn_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
...
+		/**/
 		D_GOTO(out1, rc);
 	}
 	htable->ht_vector.counter = DYNHASH_BUCKET;
 
 	/* allocate bucket */
 	D_ALLOC(bucket, sizeof(*bucket));
 	if (bucket  == NULL) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:803]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 16]
```diff
@@ -759,20 +793,23 @@ dyn_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
...
+		/**/
 		D_GOTO(out2, rc = -DER_NOMEM);
 	}
 	memset(bucket, 0, sizeof(*bucket));
 	D_MUTEX_INIT(&bucket->mtx, NULL);
 
 	/* set bucket pointer to vector */
 	for (idx = 0; idx < htable->ht_vector.counter; idx++) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:811]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 17]
```diff
@@ -782,11 +819,15 @@ dyn_hash_table_create_inplace(uint32_t feats, uint32_t bits, void *priv,
 	vec_destroy(&htable->ht_vector);
 out1:
 	if (!(feats & D_HASH_FT_NOLOCK)) {
+		/**/
 		if (feats & D_HASH_FT_MUTEX) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:823]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 18]
```diff
@@ -818,12 +859,14 @@ dyn_hash_table_traverse(struct d_hash_table *gtable, d_hash_traverse_cb_t cb,
 	for (idx = 0; idx < htable->ht_vector.counter; idx++) {
 		bucket = htable->ht_vector.data[idx];
 		if (bucket == prev) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:861]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 19]
```diff
@@ -818,12 +859,14 @@ dyn_hash_table_traverse(struct d_hash_table *gtable, d_hash_traverse_cb_t cb,
...
+			/**/
 			continue;
 		}
 		prev = bucket;
 		for (ix = 0; ix < bucket->counter; ix++) {
 			rc = cb(bucket->field[idx].record, arg);
 			if (rc != 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:868]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 20]
```diff
@@ -866,10 +909,12 @@ dyn_hash_table_destroy_inplace(struct d_hash_table *gtable, bool force)
 	for (idx = 0; idx < htable->ht_vector.counter; idx++) {
 		bucket = htable->ht_vector.data[idx];
 		if (bucket == prev) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:911]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 21]
```diff
@@ -866,10 +909,12 @@ dyn_hash_table_destroy_inplace(struct d_hash_table *gtable, bool force)
...
+			/**/
 			continue;
 		}
 		prev = bucket;
 		if (!(gtable->ht_feats & D_HASH_FT_NOLOCK)) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:916]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 22]
```diff
@@ -878,11 +923,15 @@ dyn_hash_table_destroy_inplace(struct d_hash_table *gtable, bool force)
 	htable->ht_rw_unlock(htable);
 
 	if (!(gtable->ht_feats & D_HASH_FT_NOLOCK)) {
+		/**/
 		if (gtable->ht_feats & D_HASH_FT_MUTEX) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:927]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 23]
```diff
@@ -917,11 +966,13 @@ dyn_hash_rec_find(struct d_hash_table *gtable, const void *key,
 
 	rc = find_exact_match(htable, bucket, siphash, key, ksize);
 	if (rc >= 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:968]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 24]
```diff
@@ -917,11 +966,13 @@ dyn_hash_rec_find(struct d_hash_table *gtable, const void *key,
...
+		/* get record */
 		item = bucket->field[rc].record;
 	}
 	if (item != NULL) {
 		htable->ht_ops.hop_rec_addref(gtable, item);
 		if (sip_update) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:974]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 25]
```diff
@@ -949,6 +1000,7 @@ dyn_hash_rec_find_insert(struct d_hash_table *gtable, const void *key,
 			       DH_LOOKUP_INSERT);
 	} while (rc == -DER_AGAIN);
 	if (rc == 0 && set_siphash) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1002]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 26]
```diff
@@ -973,6 +1025,7 @@ dyn_hash_rec_insert(struct d_hash_table *gtable, const void *key,
 	} while (rc == -DER_AGAIN);
 
 	if (rc == 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1027]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 27]
```diff
@@ -990,6 +1043,7 @@ dyn_hash_rec_delete(struct d_hash_table *gtable, const void *key,
 	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
 
 	if (siphash == 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1045]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 28]
```diff
@@ -1075,6 +1129,7 @@ dyn_hash_rec_decref(struct d_hash_table *gtable, dh_item_t item)
 		htable->bucket_lock(bucket);
 		do_delete(gtable, key, ksize, siphash, bucket, index);
 		if (ephemeral) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1131]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 29]
```diff
@@ -1114,6 +1169,7 @@ dyn_hash_rec_ndecref(struct d_hash_table *gtable, int count, dh_item_t item)
 		} while (--count && !zombie);
 
 		if (count != 0) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1171]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 30]
```diff
@@ -1129,6 +1185,7 @@ dyn_hash_rec_ndecref(struct d_hash_table *gtable, int count, dh_item_t item)
 		htable->bucket_lock(bucket);
 		do_delete(gtable, key, ksize, siphash, bucket, index);
 		if (ephemeral) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1187]: don't need '{}' here

File: src/gurt/dyn_hash.c [hunk 31]
```diff
@@ -1154,6 +1211,7 @@ dyn_hash_rec_first(struct d_hash_table *gtable)
 	for (idx = 0; idx < htable->ht_vector.counter; idx++) {
 		bucket = htable->ht_vector.data[idx];
 		if (bucket == prev_bucket) {
```
Comment (dmiter) [inline src/gurt/dyn_hash.c:1213]: don't need '{}' here

File: src/gurt/tests/test_gurt.c [hunk 32]
```diff
@@ -1956,6 +1954,7 @@ check_string_buffer(struct d_string_buffer_t *str_buf, int str_size,
 	assert_int_equal(str_buf->buf_size, buf_size);
 	assert_int_equal(str_buf->status, 0);
 	if (test_str != NULL) {
```
Comment (dmiter) [inline src/gurt/tests/test_gurt.c:1956]: don't need '{}' here

File: src/gurt/tests/test_gurt.c [hunk 33]
```diff
@@ -2050,6 +2049,7 @@ main(int argc, char **argv)
 
 	rc =  cmocka_run_group_tests(tests, init_tests, fini_tests);
 	if (rc != 0) {
```
Comment (dmiter) [inline src/gurt/tests/test_gurt.c:2051]: don't need '{}' here

**Review by dmiter — CHANGES_REQUESTED** [review]

**Review by dmiter — CHANGES_REQUESTED** [review]
I have an issue with this PR. github crashes regularly when try to display the files changed.

Maybe it can be re-submitted in new PR after addressing having comments.

File: src/gurt/dyn_hash.c [hunk 34]
```diff
@@ -140,6 +141,7 @@ static inline void
 vec_destroy(dh_vector_t *vec)
 {
 	if (vec->data != NULL) {
+		/* free RAM */
 		D_FREE(vec->data);
 	}
 }
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/dyn_hash.c:147]: Empty virtual functions with 02 GCC flag should be optimized out.

File: src/gurt/dyn_hash.c [hunk 35]
```diff
@@ -51,16 +51,16 @@ typedef struct dh_bucket {
 #define ROTATE(x, b) (uint64_t)(((x) << (b)) | ((x) >> (64 - (b))))
 
 #define HALF_ROUND(a, b, c, d, s, t) \
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/dyn_hash.c:53]: This part of the code is taken from open source and optimized for siphash calculation.
It doesn't compile if you try to frame it with do {}while0.

File: src/gurt/tests/test_gurt.c (hunk 33 above)
Comment (aosinskx) [reply src/gurt/tests/test_gurt.c:2051]: This code has been verified with checkpatch before commit. And it was no warning about line 2052.
I don't think the style warning do node use {} for a single line is right, but I'm not argue with the coding standard.
If {} covers more than one line it's legal. Do you really suggest to add a comment for each case like that?

File: src/gurt/dyn_hash.c [hunk 36]
```diff
@@ -132,6 +132,7 @@ vec_init(dh_vector_t *vec, unsigned char power)
 	vec->size = (size_t) (1 << power) * sizeof(void *);
 	D_ALLOC(vec->data, sizeof(*vec->data) * (1 << power));
 	if (vec->data == NULL) {
+		/* out of memory */
 		rc = -DER_NOMEM;
 	}
 	return rc;
```
[the remark this answers is not in the document]
Comment (aosinskx) [reply src/gurt/dyn_hash.c]: Disagree, memset initializes the entire structure, your proposal doesn't protect some later added member from getting uninitialized. As for this particular function: it gets called once during hash initialization.

**Comment by aosinskx** [comment]
Reopen PR due to the github access problems.

**Pull request closed without merging**